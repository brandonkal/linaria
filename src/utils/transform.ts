import path from 'path';
import * as babel from '@babel/core';
import { SourceMapGenerator } from 'source-map';
import loadOptions, { PluginOptions } from './loadOptions';
import debug from 'debug';
const log = debug('linaria:loader');
import {
  Rules,
  Replacement,
  LinariaMetadata,
  CSSIdentifiers,
} from '../babel/types';

type Result = {
  code: string;
  sourceMap: Object | null | undefined;
  cssText?: string;
  cssSourceMapText?: string;
  dependencies?: string[];
  rules?: Rules;
  replacements?: Replacement[];
  identifiers?: CSSIdentifiers;
};

type Options = {
  filename: string;
  outputFilename?: string;
  inputSourceMap?: Object;
  pluginOptions?: Partial<PluginOptions>;
};

export default function transform(code: string, options: Options): Result {
  // Check if the file contains `css`, `styled`, or injectGlobal words first
  // Otherwise we should skip transforming
  if (
    !(
      /\b(styled|css|injectGlobal)/.test(code) &&
      code.includes('@brandonkal/linaria')
    )
  ) {
    log(`skipping ${options.filename}: linaria usage not found`);
    return {
      code,
      sourceMap: options.inputSourceMap,
    };
  }

  const pluginOptions = loadOptions(options.pluginOptions);

  log(`parsing ${options.filename}`);

  // Parse the code first so babel uses user's babel config for parsing
  // We don't want to use user's config when transforming the code
  const ast = babel.parseSync(code, {
    ...(pluginOptions ? pluginOptions.babelOptions : null),
    filename: options.filename,
    caller: { name: '@brandonkal/linaria' },
  });

  log(`transforming ${options.filename}`);

  const { metadata, code: transformedCode, map } = babel.transformFromAstSync(
    ast!,
    code,
    {
      filename: options.filename,
      presets: [[require.resolve('../babel'), pluginOptions]],
      babelrc: false,
      configFile: false,
      sourceMaps: true,
      sourceFileName: options.filename,
      inputSourceMap: options.inputSourceMap,
    }
  )!;

  log(`transform complete ${options.filename}`);

  if (
    !metadata ||
    !(metadata as babel.BabelFileMetadata & { linaria: LinariaMetadata })
      .linaria
  ) {
    return {
      code,
      sourceMap: options.inputSourceMap,
    };
  }

  const {
    rules,
    replacements,
    dependencies,
  } = (metadata as babel.BabelFileMetadata & {
    linaria: LinariaMetadata;
  }).linaria;

  // Construct a CSS-ish file from the unprocessed style rules
  let cssText = buildCSS(rules);

  // When writing to a file, we need to adjust the relative paths inside url(..) expressions
  // It'll allow css-loader to resolve an imported asset properly
  if (options.outputFilename) {
    cssText = cssText.replace(
      /\b(url\()(\.[^)]+)(\))/g,
      (_, p1, p2, p3) =>
        p1 +
        // Replace asset path with new path relative to the output CSS
        path.relative(
          path.dirname(options.outputFilename!),
          // Get the absolute path to the asset from the path relative to the JS file
          path.resolve(path.dirname(options.filename), p2)
        ) +
        p3
    );
  }
  cssText += '\n';

  return {
    code: transformedCode || '',
    cssText,
    rules,
    replacements,
    dependencies,
    sourceMap: map,

    get cssSourceMapText() {
      const generator = new SourceMapGenerator({
        file: options.filename,
      });
      generator.addMapping({
        generated: {
          line: 1,
          column: 0,
        },
        original: {
          line: 1,
          column: 0,
        },
        source: options.filename,
      });
      generator.setSourceContent(options.filename, code);
      return generator.toString();
    },
  };
}

function buildCSS(rules: Rules) {
  let cssText = '';
  if (!rules) {
    return cssText;
  }
  Object.keys(rules).forEach(selector => {
    const rule = rules[selector];
    // Append new lines until we get to the start line number
    let line = cssText.split('\n').length;
    while (rule.start && line < rule.start.line) {
      cssText += '\n';
      line++;
    }
    if (!rule.isGlobal) {
      cssText += `${selector} {`;
    }
    // Append blank spaces until we get to the start column number
    const last = cssText.split('\n').pop();
    let column = last ? last.length : 0;
    while (rule.start && column < rule.start.column) {
      cssText += ' ';
      column++;
    }
    cssText += rule.cssText;
    if (!rule.isGlobal) {
      cssText += ' }';
    }
  });
  return cssText;
}
