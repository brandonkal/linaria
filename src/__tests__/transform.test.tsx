/* eslint-disable no-template-curly-in-string */

import dedent from 'dedent';
import transform from '../utils/transform';

it('rewrites a relative path in url() declarations', async () => {
  const { cssText } = await transform(
    dedent`
    import { css } from '@brandonkal/linaria';

    export const title = css\`
      background-image: url(./assets/test.jpg);
    \`;
    `,
    {
      filename: './test.js',
      outputFilename: '../.linaria-cache/test.css',
    }
  );

  expect(cssText).toMatchSnapshot();
});

it('injects global rules to the global scope', async () => {
  const { cssText } = await transform(
    dedent`
    import { injectGlobal } from '@brandonkal/linaria';

    injectGlobal\`
      /* an invalid rule */
      font-size: 10px;

      html {
        margin: 0;
      }

      body {
        color: black;
      }

    \`;
    `,
    {
      filename: './test.js',
      outputFilename: '../.linaria-cache/test.css',
    }
  );

  expect(cssText).toMatchSnapshot();
});

it('rewrites multiple relative paths in url() declarations', async () => {
  const { cssText } = await transform(
    dedent`
    import { css } from '@brandonkal/linaria';

    export const title = css\`
      @font-face {
        font-family: Test;
        src: url(./assets/font.woff2) format("woff2"), url(./assets/font.woff) format("woff");
      }
    \`;
    `,
    {
      filename: './test.js',
      outputFilename: '../.linaria-cache/test.css',
    }
  );

  expect(cssText).toMatchSnapshot();
});

it("doesn't rewrite an absolute path in url() declarations", async () => {
  const { cssText } = await transform(
    dedent`
    import { css } from '@brandonkal/linaria';

    export const title = css\`
      background-image: url(/assets/test.jpg);
    \`;
    `,
    {
      filename: './test.js',
      outputFilename: '../.linaria-cache/test.css',
    }
  );

  expect(cssText).toMatchSnapshot();
});

it('respects passed babel options', async () => {
  expect.assertions(2);
  try {
    await transform(
      dedent`
      import { css } from '@brandonkal/linaria';

      export const error = <jsx />;
      `,
      {
        filename: './test.js',
        outputFilename: '../.linaria-cache/test.css',
        pluginOptions: {
          babelOptions: {
            babelrc: false,
            configFile: false,
            presets: [['@babel/preset-env', { loose: true }]],
          },
        },
      }
    );
  } catch (e) {
    expect(e.message.includes('Unexpected token')).toBe(true);
  }

  await expect(
    transform(
      dedent`
      import { css } from '@brandonkal/linaria';

      export const error = <jsx />;
      export const title = css\`
        background-image: url(/assets/test.jpg);
      \`;
      `,
      {
        filename: './test.js',
        outputFilename: '../.linaria-cache/test.css',
        pluginOptions: {
          babelOptions: {
            babelrc: false,
            configFile: false,
            presets: [
              ['@babel/preset-env', { loose: true }],
              '@babel/preset-react',
            ],
          },
        },
      }
    )
  ).resolves.not.toThrow('Unexpected token');
});

it('handles transpiled template literals', async () => {
  expect.assertions(2);

  const result = await transform(
    dedent`
    import { css } from '@brandonkal/linaria';

    export const ok = <jsx />;
    export const title = css\`
      background-image: url(/assets/test.jpg);
    \`;
    `,
    {
      filename: './test.js',
      outputFilename: '../.linaria-cache/test.css',
      pluginOptions: {
        babelOptions: {
          babelrc: false,
          configFile: false,
          presets: [
            [
              '@babel/preset-env',
              {
                loose: true,
                targets: {
                  ie: 11,
                },
              },
            ],
            '@babel/preset-react',
          ],
        },
      },
    }
  );

  expect(result.code).toMatchSnapshot();
  expect(result.cssText).toMatchSnapshot();
});

it("doesn't throw due to duplicate preset", async () => {
  expect.assertions(1);

  await expect(
    transform(
      dedent`
      import { styled } from '../react';

      const Title = styled.h1\` color: blue; \`;

      const Article = styled.article\`
        ${'${Title}'} {
          font-size: 16px;
        }
      \`;
      `,
      {
        filename: './test.js',
        outputFilename: '../.linaria-cache/test.css',
        pluginOptions: {
          babelOptions: {
            babelrc: false,
            configFile: false,
            presets: [require.resolve('../babel')],
            plugins: [
              require.resolve('@babel/plugin-transform-modules-commonjs'),
            ],
          },
        },
      }
    )
  ).resolves.not.toThrow('Duplicate plugin/preset detected');
});
