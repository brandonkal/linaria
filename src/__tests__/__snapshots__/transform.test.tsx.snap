// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't rewrite an absolute path in url() declarations 1`] = `
"

.title_tpyglzj {
  background-image: url(/assets/test.jpg);
 }
"
`;

exports[`injects global rules to the global scope 1`] = `
"


  /* an invalid rule */
  font-size: 10px;

  html {
    margin: 0;
  }

  body {
    color: black;
  }


"
`;

exports[`rewrites a relative path in url() declarations 1`] = `
"

.title_tpyglzj {
  background-image: url(../linaria/assets/test.jpg);
 }
"
`;

exports[`rewrites multiple relative paths in url() declarations 1`] = `
"

.title_tpyglzj {
  @font-face {
    font-family: Test;
    src: url(../linaria/assets/font.woff2) format(\\"woff2\\"), url(../linaria/assets/font.woff) format(\\"woff\\");
  }
 }
"
`;
